app/src/main/java/com/nativo/sampleapp/ViewAdapter/GridViewAdapter.kt

```diff
@@ -0,0 +1,165 @@
+
+    companion object {
+        const val ITEM_COUNT = 30
+    }
+
+    var integerList: MutableList<Int> = ArrayList()
+    var initialNativoRequest = true
+
+    init {
+        // Nativo init
+        initSectionWithAdapter(this, SECTION_URL, context)
+
+        for (i in 0..ITEM_COUNT) {
+            integerList.add(i)
+        }
+    }
+
+    override fun getCount(): Int {
+        return integerList.size
+    }
+
+    override fun getView(i: Int, view: View?, viewGroup: ViewGroup): View {
+        // Publisher article view
+        val convertView: View
+        val holder: ItemViewHolder
+
+        if (view == null) {
+            convertView = LayoutInflater.from(viewGroup.context)
+                .inflate(R.layout.publisher_article, viewGroup, false)
+            holder = ItemViewHolder(convertView)
+            convertView.tag = holder
+        } else {
+            convertView = view
+            holder = view.tag as ItemViewHolder
+        }
+
+        holder.bind(integerList[i], i)
+        return convertView
+    }
+
+    inner class ItemViewHolder(private val container: View) {
+        private val articleImage: ImageView? = container.findViewById(R.id.article_image)
+        private val articleTitle: TextView? = container.findViewById(R.id.article_title)
+        private val articleAuthor: TextView? = container.findViewById(R.id.article_description)
+        private val articleSponsor: ImageView? = container.findViewById(R.id.sponsored_ad_indicator)
+
+        fun bind(item: Int, position: Int) {
+            articleImage?.setImageResource(R.drawable.newsimage)
+            articleSponsor?.visibility = View.INVISIBLE
+            articleAuthor?.setText(R.string.sample_author)
+            articleTitle?.setText(R.string.sample_title)
+
+            container.setOnClickListener {
+                context.startActivity(
+                    Intent(
+                        Intent.ACTION_VIEW,
+                        Uri.parse(CLICK_OUT_URL)
+                    )
+                )
+            }
+
+            if (shouldShowPlaceAd(item)) {
+                val isAdContentAvailable =
+                    placeAdInView(container, gridView, SECTION_URL, position, null)
+                Log.w(NtvTAG, "isAdContentAvailable = $isAdContentAvailable")
+            }
+        }
+
+        private fun shouldShowPlaceAd(item: Int): Boolean {
+            return item % 3 == 1
+        }
+    }
+
+    override fun didReceiveAd(didGetFill: Boolean, inSection: String) {
+        Log.d(NtvTAG, "Did receive ad: $didGetFill")
+
+        if (didGetFill && initialNativoRequest) {
+            Log.w(NtvTAG, "Needs Reload Everything")
+            notifyDataSetChanged()
+        }
+
+        initialNativoRequest = false
+    }
+
+    override fun didAssignAdToLocation(
+        location: Int,
+        adData: NtvAdData,
+        inSection: String,
+        container: ViewGroup
+    ) {
+        Log.d(NtvTAG, "didAssignAdToLocation: $location")
```

Comment:
When an ad gets assigned to a location, you should add an extra row to your datasource(integerList) and then inject that row here